### ISSUE TITLE
Various repositories for Spring integration and old documentation - which one to use and which version?

### ISSUE THREAD

**Issue by DanailMinchev** [auth0/auth0-spring-security-api#1]
Hi @arcseldon ,

I am trying to integrate Auth0 with Spring Boot and SPA (Angular 2 for the website and Ionic 2 for the mobile app).

I am confused which repository is best to use.

I found **4** repositories ( https://github.com/auth0?utf8=%E2%9C%93&query=spring ):

**1. https://github.com/auth0/spring-security-auth0 (forked from ddtxra/spring-security-auth0)**

Samples:
- https://github.com/auth0/spring-security-auth0/tree/master/examples/spring-boot-api-example

Pros:
- This is the only one repository which is available in Maven: http://mvnrepository.com/artifact/com.auth0/spring-security-auth0
- It has documentation for it: https://auth0.com/docs/server-apis/java-spring-security

Cons:
- You can specify only one secure url via _auth0.securedRoute_ property:

I want to use `protected void configure(HttpSecurity http) throws Exception` method  in `WebSecurityConfigurerAdapter` to authorize request paths with roles and other config (like in regular Spring Security / Spring Security OAuth application configuration)

- Not well integrated with Spring Boot

**2. https://github.com/auth0/auth0-spring-mvc**

Samples:
- https://github.com/auth0-samples/auth0-spring-mvc-sample

The official documentation is **404**:
https://auth0.com/docs/server-platforms/java-spring-mvc

Description of the repository:

> A modern Java Spring library that allows you to use Auth0 with Java Spring for server-side MVC web apps. Leverages Spring Boot dependencies. Validates the JWT from Auth0 in every API call to assert authentication according to configuration. If your application only needs secured endpoints and the ability to programmatically work with a Principal object for GrantedAuthority checks this library is a good fit. However, if you are already using Java Spring and wish to leverage fully Java Spring Security - with powerful support for Security Annotations, Security JSTL Tag libraries, Fine-grained Annotation level method security and URL endpoint security at the Role / Group level - then see this project **Auth0 Spring Security MVC** and associated sample **Auth0 Spring Security MVC Sample**

From the description of the repository it is clear that if I want Spring Boot + Spring Security I should use the other repo **Auth0 Spring Security MVC** (https://github.com/auth0/auth0-spring-security-mvc) and associated sample **Auth0 Spring Security MVC Sample**

**3. https://github.com/auth0/auth0-spring-security-mvc**

Samples:
- https://github.com/auth0-samples/auth0-spring-security-mvc-sample

The official documentation is **404**:
https://auth0.com/docs/server-apis/java-spring-security-mvc

This is the repository suggested by description in https://github.com/auth0/auth0-spring-mvc if I want to  use Spring Boot + Spring Security + Java Configuration

But, it is not available in Maven and has no official documentation: https://auth0.com/docs/server-apis/java-spring-security-mvc
Also what is the difference between this repository and the next one?

**4. https://github.com/auth0/auth0-spring-security-api**

Samples:
- https://github.com/auth0-samples/auth0-spring-security-api-sample

The official documentation is **404**:
https://auth0.com/docs/server-apis/java-spring-security-api

This repository is similar to 3rd one, also I found your comment on it https://github.com/auth0/auth0-spring-security-mvc/issues/1#issuecomment-224158819 and associated issue similar to which I have here: https://github.com/auth0/auth0-spring-security-mvc/issues/1

What is the difference between option 3 and 4 ?
- https://github.com/auth0/auth0-spring-security-mvc/tree/master/src/main/java/com/auth0/spring/security/mvc
- https://github.com/auth0/auth0-spring-security-api/tree/master/src/main/java/com/auth0/spring/security/api

----------

Please clarify which repository should I use? Why there are so many repositories? Some of them will be removed/deprecated (if yes which one)?

Currently it's really confusing...

I was hoping to start with the official documentation https://auth0.com/docs/server-apis/java-spring-security for the only one available Maven lib, but it is not generating the project seed:
```
{"cause":{"errno":-2,"code":"ENOENT","syscall":"open","path":"/tmp/githubdownload116511-27734-kdatwq/spring-security-auth0-master/examples/api-example/src/main/resources/auth0.properties"},"isOperational":true,"errno":-2,"code":"ENOENT","syscall":"open","path":"/tmp/githubdownload116511-27734-kdatwq/spring-security-auth0-master/examples/api-example/src/main/resources/auth0.properties"}
```
Anyway, this is another issue.

Thank you

**Comment by arcseldon** [comment]
@DanailMinchev - Great questions, and I really appreciate the detailed information with links etc above. Fully appreciate your source of confusion - we have just begin a large drive to refresh our Java offerings and embrace the convenience of Spring Boot and integration with Spring Security for both API and MVC scenarios too (but also offer plain Java / Java Spring standalone for those that really need it).

You just happened to approach us on the very weekend all this is going on and caught the libraries and documentation in a state of flux.

To set the record straight, here are our offerings (none of which have latest versions in Maven yet - I promise you they shall be published Monday 13th June 2016 - today for me in Japan) - however our publication process involves QA so I cannot just go ahead and publish now. However, for working locally, just `git clone <library_repo>` - then do `mvn clean install` - you will have a local copy of the repo you need and should be able to reference that from the sample you are working against.

**Current Libraries**

[https://github.com/auth0/auth0-servlet](https://github.com/auth0/auth0-servlet) - no frills, servlet based solution - mostly available for legacy integration needs with older Java style web apps. 

[https://github.com/auth0/auth0-spring-mvc](https://github.com/auth0/auth0-spring-mvc) - A modern Java Spring library that allows you to use Auth0 with Java Spring for server-side MVC web apps. This is the right choice if you are using Spring / Spring Boot but don't want to use Spring Security - perhaps your app is using / shall integrate with alternate Security strategy.

[https://github.com/auth0/auth0-spring-security-mvc](https://github.com/auth0/auth0-spring-security-mvc) - A modern Java Spring library that allows you to use Auth0 with Spring Security for server-side MVC web apps. This provides full integration with Spring Integration.

ALL the above basically adopt the `Oauth2 / OIDC Authorization Code Grant` flow in which authentication results in a callback to the server-side application with a `code`. This is then exchanged for `id_token` (JWT Token) and `access_token` on the server-side (as part of the callback), and once the tokens have been received by the application, then UserProfile information can also be retrieved (via request with a valid Token). 

For your scenario - using `Single Page Applications` (which shall use `Oauth2 / OIDC  Implicit Grant Flow` in which authentication results directly in receiving a JWT Token you should use the following library:

[https://github.com/auth0/auth0-spring-security-api](https://github.com/auth0/auth0-spring-security-api) - A modern Java Spring library that allows you to use Auth0 with Spring Security. Leverages Spring Boot dependencies. Validates the JWT from Auth0 in every API call to assert authentication according to configuration. This library would be suitable for headless APIs and SPA (single page application) backend end server scenarios. Sample here: [https://github.com/auth0-samples/auth0-spring-security-api-sample](https://github.com/auth0-samples/auth0-spring-security-api-sample).

I haven't had time to provide a sample with SPA yet, however you have two usage scenarios. You could either co-locate the SPA inside the same Spring Boot app from which you use this library (drop your SPA code under `/src/main/resources` as you would for any Spring Boot application - and then the SPA talks directly to its web-server (Spring Boot) using JWT authorization on all requests to secured endpoints as defined by your AppConfig ( just augment [here](https://github.com/auth0-samples/auth0-spring-security-api-sample/blob/master/src/main/java/com/auth0/example/AppConfig.java#L96) ). 

@DanailMinchev - I shall try to get an Angular 1.x SPA sample written in the next few hours to clearly show users like yourself how to integrate a SPA solution with the new libraries. I am just half-way through providing a separate sample for SSO (customer request last Friday so shall do so once that is published). I shall also continue to complete all the README information to make all the above clearer, and bear in mind we have a dedicated Content Team (docs) that shall ensure our Documentation (separate from README info) is all up to date and reflects the code changes made over past 72 hours.

Also, realise there is going to be some consolidation and internal refactoring shortly to pull out any duplication across our libraries and house these as separate `auth0-java-commons` jar files. For users currently perusing the source code, it does look like for example the Spring API and MVC libraries are almost identical. They are very similar but placing the shared code into a support Jar file would make it much more obvious where the crucial differences are between the libraries.

With regards to what is effectively end of life. Once our latest versions of the above libraries are published today (maven / gradle), then the following are effectively dead:

Any version of `auth0-servlet` less than version 3.
Any version of `spring-security-auth0`.

We'll leave those published for legacy user support but would not recommend further adoption since the new libraries are the future of our Java Offering.

Ultimately, what is particularly confusing for our scenario as things stand, is that a SPA requires view endpoints defined (as you alluded to above) in `protected void configure(HttpSecurity http)` but you need the behaviour of an API, not server-side MVC using authorization code grant flow. In that regards, this [config](https://github.com/auth0-samples/auth0-spring-security-mvc-sample/blob/master/src/main/java/com/auth0/example/AppConfig.java#L86) may be closer to what you need... also take a look at the layout of the `src/main/resources` area of that sample since it contains static resources for view presentation.  Like I say, i shall get a SPA app sample written so this is all much clearer for users with similar requirements to yourself.

FYI only:

[auth0-java](https://github.com/auth0/auth0-java) - this is a Client library to make calls to our Authentication and Management APIs.

[java-jwt](https://github.com/auth0/java-jwt) - this is used for performing JWT Token related actions - JWT creation, verification etc.

Make sense now?  Please feel free to contact me directly if you have questions or get stuck (richard.seldon@auth0.com) - or continue to post comments on this issue here too. Do what I can to ensure you get your app up and running asap. Again, thanks for the valued feedback.

**Comment by DanailMinchev** [comment]
@arcseldon 

Thank you for your reply! It is much clearer now and I am very happy to know that it all be finished/published soon. I really appreciate your hard work!

According to `Single Page Applications` I think the best example for me will be `Angular 2` app which is not part of `Spring Boot` (not included in `/src/main/resources`) - it can be `Node.js` app which I can just `npm install` and `npm start` using local server (like https://www.npmjs.com/package/lite-server) and which is fully detached from the API server / Spring Boot and authenticates using `JWT`.
Another `SPA` example will be `Ionic 2` app like this one: https://github.com/auth0/auth0-ionic2

Anyway, I think finishing `Spring Boot` + cURL examples (which can be used to understand how to construct for example different `http` requests with Angular's `http` service .. etc - `cURL` is nice abstraction and other users can integrate their frameworks once they have `cURL` examples) + fixing the docs is the first priority. I can use existing `SPA` examples with the newer `Spring Boot` integration.

Just one more question:

I will use https://github.com/auth0/auth0-spring-security-api and https://github.com/auth0-samples/auth0-spring-security-api-sample for building the API server with `Spring Boot` and some SPA like `Angular 2`. The SPA will authenticate with JWT tokens for each request.
You said that you will refactor https://github.com/auth0/auth0-spring-security-api and extract the common behaviour in separate lib `auth0-java-commons`.
My question: what is the estimated time for this refactoring? Should I start the project as this one https://github.com/auth0-samples/auth0-spring-security-api-sample now or it will be changed in few days anyway (if it's going to be in few weeks - I will use the current project, but if it will be updated in few days with major refactoring I will wait)?
In short: could you tell me when stable version of https://github.com/auth0/auth0-spring-security-api + docs/examples will be available?

Thank you very much!

**Comment by arcseldon** [comment]
@DanailMinchev - most welcome. Regarding cURL examples - please see [here](https://github.com/auth0-samples/auth0-spring-security-api-sample#test-the-api) - there is also a [postman collection](https://github.com/auth0-samples/auth0-spring-security-api-sample/tree/master/postman) for the endpoints in the that particular sample - you can import into [postman](https://www.getpostman.com/) if you prefer to test your API using that app (my personal recommendation). This should show you how to do the same for testing your own particular SPA app endpoints.

The libraries above are now relatively stable - you can safely start using any of the above. The refactoring I mentioned would be internal and not change behaviour). It would result in pulling common library code into a dependency jar that the library itself (not you, the end user application) would reference. So from your perspective, nothing would change - you'd only need to update your version of the maven library if you wished to pull in the version of our library with the internally refactored library dependency (optional).

Please go ahead and start building out your application - I'd certainly appreciate any feedback you have from doing so, and if you wish to contribute directly please feel free to share a link to your own app on github etc - could provide the inspiration for another of our samples :) - again if you are writing something private / commercial then no worries. What interests me most is how easy you find the libraries to work with, any rough edges or suggestions on how to make things better.

One idea I have already, but am waiting until things stabilise (published to maven, docs have caught up and we've received real-world feedback from users such as yourself) before implementing is consolidation of jar dependences by placing them inside `auth0-spring-boot-starter` dependencies. So instead of having to reference several dependencies in your maven `pom.xml` (or gradle build file), you'll just have to reference a single `auth0-boot-starter` (there may be a few types depending on type of application you are building and whether you want spring security) and it will transitively pull in all the required dependencies for you. That way, the dependency management gets simplified to be trivial. As stated already, am reluctant to do this until we've ironed out any creases in the current offerings and have a clearer understanding on what is needed in that regards.

**Comment by arcseldon** [comment]
@DanailMinchev - the java libraries above have now been published to maven,  and samples referenced above are updated to accurately point at correct versions of the published libraries.

Hope to have a sample for SPA app scenarios later in the week. The readmes in each of the libraries and in particular the samples are mostly done - however, the site documentation and quick starts are still catching up.

**Comment by DanailMinchev** [comment]
@arcseldon 

Great! Thank you for your updates!
I think the original request is resolved and I am closing this issue.

Cheers!
